firebase-vertexai/src/main/kotlin/com/google/firebase/vertexai/type/AudioHelper.kt

```diff
@@ -17,125 +17,160 @@
+   * Returns an empty flow if this [AudioHelper] has been [released][release].
+   */
+  fun listenToRecording(): Flow<ByteArray> {
+    if (released) return emptyFlow()
+
+    resumeRecording()
+
+    return recorder.readAsFlow()
   }
-  @RequiresPermission(Manifest.permission.RECORD_AUDIO)
-  fun startRecording(): Flow<ByteArray> {
-
-    val bufferSize =
-      AudioRecord.getMinBufferSize(
-        16000,
-        AudioFormat.CHANNEL_IN_MONO,
-        AudioFormat.ENCODING_PCM_16BIT
-      )
-    if (
-      bufferSize == AudioRecord.ERROR ||
-        bufferSize == AudioRecord.ERROR_BAD_VALUE ||
-        bufferSize <= 0
-    ) {
-      throw AudioRecordInitializationFailedException(
-        "Audio Record buffer size is invalid (${bufferSize})"
-      )
-    }
-    audioRecord =
-      AudioRecord(
-        MediaRecorder.AudioSource.VOICE_COMMUNICATION,
-        16000,
-        AudioFormat.CHANNEL_IN_MONO,
-        AudioFormat.ENCODING_PCM_16BIT,
-        bufferSize
-      )
-    if (audioRecord.state != AudioRecord.STATE_INITIALIZED) {
-      throw AudioRecordInitializationFailedException(
-        "Audio Record initialization has failed. State: ${audioRecord.state}"
-      )
-    }
-    if (AcousticEchoCanceler.isAvailable()) {
-      val echoCanceler = AcousticEchoCanceler.create(audioRecord.audioSessionId)
-      echoCanceler?.enabled = true
-    }
 
-    audioRecord.startRecording()
-
-    return flow {
-      val buffer = ByteArray(bufferSize)
-      while (!stopRecording) {
-        if (audioRecord.recordingState != AudioRecord.RECORDSTATE_RECORDING) {
-          buffer.fill(0x00)
-          continue
-        }
-        try {
-          val bytesRead = audioRecord.read(buffer, 0, buffer.size)
-          if (bytesRead > 0) {
-            emit(buffer.copyOf(bytesRead))
-          }
-        } catch (_: Exception) {}
+  companion object {
+    /**
+     * Creates an instance of [AudioHelper] with the track and record initialized.
+     *
+     * A separate build method is necessary so that we can properly propagate the required manifest
+     * permission, and throw exceptions when needed.
+     *
+     * It also makes it easier to read, since the long initialization is separate from the
+     * constructor.
+     */
+    @RequiresPermission(Manifest.permission.RECORD_AUDIO)
+    fun build(): AudioHelper {
+      val playbackTrack =
+        AudioTrack(
+          AudioAttributes.Builder().setUsage(AudioAttributes.USAGE_VOICE_COMMUNICATION).build(),
+          AudioFormat.Builder()
+            .setSampleRate(24000)
+            .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
+            .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
+            .build(),
+          AudioTrack.getMinBufferSize(
+            24000,
+            AudioFormat.CHANNEL_OUT_MONO,
+            AudioFormat.ENCODING_PCM_16BIT
+          ),
+          AudioTrack.MODE_STREAM,
+          AudioManager.AUDIO_SESSION_ID_GENERATE
+        )
+
+      playbackTrack.play()
```

Comment:
when building an audioHelper, is it expected that we start playing too?

Reply:
Yeah, since we're streaming the audio data into it.

Reply:
TBH I wouldn't expect a `build` method to have side effects like start playing audio. 

We can either have another method to start playing, or, in the `playAudio` method check the state (once) and start playing if not.

Reply:
SGTM, done